activation/poet.go

```diff
@@ -412,21 +425,48 @@ func NewPoetServiceWithClient(
 	logger *zap.Logger,
 	opts ...PoetServiceOpt,
 ) *poetService {
-	poetClient := &poetService{
+	service := &poetService{
 		db:                 db,
 		logger:             logger,
 		client:             client,
 		requestTimeout:     cfg.RequestTimeout,
-		certifierInfoCache: cachedData[*certifierInfo]{ttl: cfg.CertifierInfoCacheTTL},
+		certifierInfoCache: cachedData[*types.CertifierInfo]{ttl: cfg.CertifierInfoCacheTTL},
 		powParamsCache:     cachedData[*PoetPowParams]{ttl: cfg.PowParamsCacheTTL},
 		proofMembers:       make(map[string][]types.Hash32, 1),
+		expectedPhaseShift: cfg.PhaseShift,
+		fetchedPhaseShift:  &atomic.Duration{},
 	}
-
 	for _, opt := range opts {
-		opt(poetClient)
+		opt(service)
 	}
 
-	return poetClient
+	err := service.verifyPhaseShiftConfiguration(context.Background())
+	if err != nil {
+		if errors.Is(err, ErrIncompatiblePhaseShift) {
+			logger.Panic("failed to create poet service",
+				zap.Error(err),
+				zap.String("poet", client.Address()))
+		}
+		logger.Warn("failed to fetch poet phase shift",
+			zap.Error(err),
+			zap.String("poet", client.Address()))
+	}
+	return service
+}
+
+func (c *poetService) verifyPhaseShiftConfiguration(ctx context.Context) error {
+	if c.fetchedPhaseShift.Load() != 0 {
+		return nil
+	}
+	resp, err := c.client.Info(ctx)
+	if err != nil {
+		return err
+	} else if resp.PhaseShift != c.expectedPhaseShift {
+		return ErrIncompatiblePhaseShift
+	}
+
+	c.fetchedPhaseShift.Store(resp.PhaseShift)
+	return nil
 }
```

Comment:
Also I just noticed, an atomic value here is probably not the right way to go. Although `fetchedPhaseShift` can now be accessed safely from multiple routines, the method `verifyPhaseShiftConfiguration` is not protected with the atomic value. If it is called from multiple go routines before `fetchedPhaseShift` has been set, all those calls will trigger an HTTP request to PoET, although it would be better if only the first did.

I think the right way to go here is not an atomic value but instead a field protected by a mutex.